### ISSUE TITLE
Large amounts of files missing

### ISSUE THREAD

**Issue by boredazfcuk** [icloud-photos-downloader/icloud_photos_downloader#99]
I've installed iCloud control panel in a Windows 7 VM and kicked off a download of all files and to separate them into folders by year.
I've set icloud_photos_downloader to download using --folder-structure={:%Y} so that it also downloads to different folders based on the year.

I would expect to see both folders identical contents. Instead I see a large amounts missing files, on both sides: https://i.imgur.com/VZDRoQp.png

**Comment by boredazfcuk** [comment]
I've cleaned everything out and re-run the downloads to test again... It's more complete as in ti's only missing a handful of files this time around, but it's still not got them all. Oddly, it has downloaded some that the iCloud downloader hasn't:

https://i.imgur.com/eWBUTPk.png
https://i.imgur.com/CnEVppt.png

**Comment by ndbroadbent** [comment]
Hi, sorry I didn't respond to this issue earlier. I've made a few changes to the code so that it now uses the "fingerprint" if a filename was missing. That might solve some of the missing files.

But yes this is something I do need to check. I think I will use the Photos app to download all of my photos to my Mac, and then compare the results with the script.

**Comment by realtek2017** [comment]
Hi,

I've spent months and months on this issue, not just with this tool but others (including iCloud for windows).

The issue is due to duplicates that get stored in iCloud. I have 43,000 photos and after exporting using iCloud, Photos for Mac or this tool I end up short of 3,000.

This photos downloader python script allowed me to identify the issue and I made a lot of changes to it to test.

I believe the duplicates happen mostly because the device does not get confirmation the photo uploaded and therefore does it again so duplicates appear in iCloud.

But the issue really is a problem when exporting, because tools store the same duplicate photo on the local filesystem in the same folder and obviously cannot have the same filename.

I modified the code to actually add a random number at the end if the file already exists. I now have a much better photo count between this and iCloud.

Obviously we do not care about duplicates, but I would be more comfortable comparing these photos manually than assuming its a duplicate because the filename is the same.

My modified code only works on a single run, if the script ends or exits you need to delete all the photos and start again but I did it to troubleshoot the issue.

Not sure how this can be done for an enhancement - my only suggestion could be to have a new parameter "allow-duplicates" and add the current photo number to the end of each filename?

**Comment by AndreyNikiforov** [comment]
https://github.com/ndbroadbent/icloud_photos_downloader/pull/160 uses size in addition to names when comparing assets (and created duplicates with different names). Should solve the problem of duplicate names (in most/many cases).

**Comment by menkej** [comment]
Could you please recheck if this issue is still valid with the current version (and changes for the handling of duplicates)? @boredazfcuk @realtek2017

**Comment by boredazfcuk** [comment]
I don't actually run the Windows iCloud control panel to run a comparison any more. I'll need to build a VM and install it to test. Struggling for time at the minute due to work commitments, hopefully I'll be able to look at it later in the week.

I think I'm still seeing issues with duplicates in my Docker container however. I have a function that creates a list of files on the server using the `--only-print-filenames` option. It then checks if each file exists on disk and if it does, deletes the entry from the list. At the end of the function, if there are still lines left in the list, my script knows there are files that need to be downloaded and moves to the download phase. This prevents it from attempting to download files every single time it runs. I did this to help combat the Apple API throttling issue.

This is the function in question:
https://github.com/boredazfcuk/docker-icloudpd/blob/f5475afa2006079041e37f73ad06c3161f6eeacc/sync-icloud.sh#L272

Looking at the logs for the three Apple IDs that I sync, it seems that it's only detecting duplicate .MOV files. I have noticed the log has entries for some pictures that have a notification about being de-duplicated... Maybe the de-dupe code only applies to HEIC/JPEG/PNG files?

**Comment by AndreyNikiforov** [comment]
@boredazfcuk de-dup works on all files types. One caveat though: if you choose to set exif dates for images, then file size will be different and subsequent runs may identify existing file as a duplicate. Example:

0. you have only one file abc.jpg in your collection with the size 1024bytes without exif
1. first run: downloaded abc.jpg 1024bytes; updated its exif and now size on disk is 1100bytes.
2. second run: tools checks that file abc.jpg exists, but its size is different (1100bytes) than in iCloud (1024bytes), so tool downloads same file again with name abc-1024.jpg and changes its exif; size of the abc-1024.jpg on disk is 1100bytes now too.
3. third run: abc.jpg exists, so tool checks abc-1024.jpg, which exists, to the tool considers as existing and passes to the next (existing in the case of this example)

**Comment by boredazfcuk** [comment]
Maybe the issue I see is unrelated to de-dupe then... but I find that the script always detects certain files need downloading. This is the log output from my container: https://pastebin.com/QB1zy4Z9

As you can see, it detects the same 10 files requiring download, even though they already exist on the disk.

I've just logged on to iCloud.com and can see that these 10 files are actually ten "live photos" that are in the photo stream... I only have 10 live photos too, so I can't imagine that's just a coincidence.

**Comment by boredazfcuk** [comment]
So I've built a Win10 VM this morning, installed iCloud Control Panel and synchronised my photo stream. It downloaded 839 images.

I then created new icloudpd container, kicked off a download, and it's downloaded 838 images.

For reference, my phone's "recents" album also shows a file count of 838.

I've created two file lists and used WinMerge to compare the differences in the file lists. This is my conclusion on the file list comparison:

iCloudPD downloads HEIC whereas iCloudCP downloads JPEG files. A quick search/replace meant that there were only 13 lines with differences.

10 of those differences were photos taken in my Telegram app (which prefixes files with camphoto_ and has a habit of duplicating file names). These 10 file were successfully de-duped. They were showing as camphoto_*(1).jpg camphoto_*(2).jpg etc in my Photo Stream, but as camphoto_*-<file_size>.jpg... All good.

That left 4 other differences. The first was a file that iCloudPD had downloaded as:
```
IMG_1231-2545351.HEIC
IMG_1231.HEIC
```
However, only a single file IMG_1231.JPEG exists in iCloudCP,

I've attempted to open the IMG_1231.HEIC file and it is appears corrupt. Possibly, the file has failed to download correctly (my first download run did crash midway), so the second run has then de-duped the file, rather than overwrite the corrupt file.

The final two differences were two files that existed in my Photo Stream that iCloudPD just doesn't detect:
```
IMG_0924.JPG
IMG_0925.JPG
```
Bizarrely, neither of these files can be seen when looking at my photos using icloud.com, they cannot be seen on my iPhone either. They are not in 'recently deleted' and I don't use named albums (apart from the automatically created WhatsApp album). Both of these pictures are non-photos (possibly taken when I was fumbling my phone out of my pocket or something) so I presume they were deleted long ago, but are still somehow stuck in the system.

**Comment by menkej** [comment]
> This is far more accurate than what I was originally finding when I opened this ticket. I think the de-dupe feature is what has (more or less) corrected the disparity.

Thanks for your detailed analysis! Its really weird with these two files... But from how I read your post you think the program behaves correctly now?

Do you have any more details regrading the crashed download?

**Comment by boredazfcuk** [comment]
The bizarre thing is that the files now appear in my Photos app. They absolutely, definitely weren't were not there to begin with. They are two additional pictures from a "shoot" which had 5 pictures. I now have 7 pictures in that "shoot"

I didn't note the error unfortunately. I was running a dev version of my container, in which I'd inadvertently removed some Python dependencies. I moved from specifying all the Python dependencies in my Dockerfile to installing them using the requirements.txt file from icloud_photos_downloader. This meant I was no longer installing:
```
certifi pytz tzlocal six chardet idna urllib3 future keyrings.alt==1.0 keyring==8.0
```
I readded these, rebuilt my container, and it hasn't crashed in any of the other tests I've done tonight..

**Comment by AndreyNikiforov** [comment]
> I've attempted to open the IMG_1231.HEIC file and it is appears corrupt. Possibly, the file has failed to download correctly (my first download run did crash midway), so the second run has then de-duped the file, rather than overwrite the corrupt file.
> 

@boredazfcuk icloudpd 1.6.1+ downloads file to temp one (.part suffix) and then renames to final destination only once downloading completes successfully. I cannot come up to the scenario when file will appear corrupted as a result of program termination.

**Comment by AndreyNikiforov** [comment]
> I didn't note the error unfortunately. I was running a dev version of my container, in which I'd inadvertently removed some Python dependencies. I moved from specifying all the Python dependencies in my Dockerfile to installing them using the requirements.txt file from icloud_photos_downloader. This meant I was no longer installing:
> 
> ```
> certifi pytz tzlocal six chardet idna urllib3 future keyrings.alt==1.0 keyring==8.0
> ```
> 
> I readded these, rebuilt my container, and it hasn't crashed in any of the other tests I've done tonight..

`Dockerfile` in icloudpd builds container from `python` image. I found at least one case (albumNameEnc error, I haven't followed up yet) when alpine-based image was not behaving, while python-based one worked. See if that python-based fat image will be better for troubleshooting functionality of icloudpd itself.

**Comment by boredazfcuk** [comment]
> @boredazfcuk icloudpd 1.6.1+ downloads file to temp one (.part suffix) and then renames to final destination only once downloading completes successfully. I cannot come up to the scenario when file will appear corrupted as a result of program termination.

I deleted the corrupt file, re-ran the script, and the copy it downloaded in the 2nd run was fine. I'll just put it down to a transient issue.

> `Dockerfile` in icloudpd builds container from `python` image. I found at least one case (albumNameEnc error, I haven't followed up yet) when alpine-based image was not behaving, while python-based one worked. See if that python-based fat image will be better for troubleshooting functionality of icloudpd itself.

Odd, I've not experienced any issues with the Alpine images I've been maintaining (since 3.10). My container notifies me of downloads, deleted files & non-zero exit codes, plus I have three Apple IDs that I sync. So I'm fairly certain that if it were to have an error, it wouldn't be missed and I'd pick up on it fairly quickly...

Someone did raise the albumNameEnc error against my container (https://github.com/boredazfcuk/docker-icloudpd/issues/21), and I was seeing the issue myself for a short while, so I just created a diff from the patch and built my Dockerfile with that until the PR was merged. The albumNameEnc patch has been working for me since early September.

**Comment by AndreyNikiforov** [comment]
Based on the discussion, original issue (missing files) has been resolved with 1.6.1 version of the icloudpd. Closing.
